### ISSUE TITLE
Throw error for softmax_cross_entropy_with_integer_labels when sizes aren't correct

### ISSUE THREAD

**Issue by TugdualKerjan** [google-deepmind/optax#1185]
When calculating softmax_cross_entropy_with_integer_labels, an error should be thrown if the `labels` array contains a value that is larger than the `num_classes` in the logits, avoiding a silent NaN.

```python
import optax
import jax.numpy as jnp

logits = jnp.array([[0.2, 0.1, 0.4, 0.6]])
labels = jnp.array([4])
cross_entropy = optax.softmax_cross_entropy_with_integer_labels(logits, labels)
```

`cross_entropy = Array([nan], dtype=float32)`

**Comment by fabianp** [comment]
I agree, would you be willing to contribute a patch for this?

**Comment by carlosgmartin** [comment]
Since the contents of `labels` are dynamic (unlike e.g. its shape and dtype), how would this work under tracing? As [this comment](https://github.com/jax-ml/jax/issues/4257#issuecomment-690844567) says, "There's essentially no runtime error mechanism in XLA".

**Comment by rdyro** [comment]
This is definitely a problem, but like @carlosgmartin is pointing out, it's tricky in XLA because of potential compiler fusions.

Perhaps we should point users to [https://jax.readthedocs.io/en/latest/debugging/flags.html](https://jax.readthedocs.io/en/latest/debugging/flags.html) instead of applying fixes in optax?

Thanks for catching this @TugdualKerjan

**Comment by rdyro** [comment]
A possible workaround is to use the `jax.debug` module. It will have a performance impact, but works under jit.

```python
import jax
from jax import numpy as jnp
import optax


@jax.jit
def check_labels(labels, num_classes):
  def _check_labels():
    raise RuntimeError("Labels are not in [0, num_classes)")
    return 0

  return jax.lax.cond(
      jnp.all((labels < num_classes) & (labels >= 0)),
      lambda: None,
      lambda: jax.debug.callback(_check_labels),
  )


logits = jnp.array([[0.2, 0.1, 0.4, 0.6]])
labels = jnp.array([4])
cross_entropy = optax.softmax_cross_entropy_with_integer_labels(logits, labels)
check_labels(labels, num_classes=4)
print(cross_entropy)

```

**Comment by fabianp** [comment]
thanks @rdyro and @carlosgmartin , I agree with you, I hadn't thought about the JITting implications

**Comment by carlosgmartin** [comment]
Given the above, should this be closed?
